File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/LmsFeatureModelsCourseTranslations.kt
package com.lextures.android.core.lms

import kotlinx.serialization.Serializable

@Serializable
data class TranslationCoverage(
    val targetLocale: String = "",
    val totalItems: Int = 0,
    val translatedItems: Int = 0,
    val percent: Double = 0.0,
    val untranslated: List<TranslatableCourseItem> = emptyList(),
) {
    val percentInt: Int get() = percent.toInt().let {
        // Prefer rounded whole percent for display.
        kotlin.math.round(percent).toInt()
    }
}

@Serializable
data class TranslatableCourseItem(
    val itemId: String = "",
    val itemType: String = "",
    val title: String = "",
    val body: String = "",
    val hasPublished: Boolean? = null,
    val hasDraft: Boolean? = null,
)

@Serializable
data class CourseTranslationListResponse(
    val items: List<CourseTranslationListItem> = emptyList(),
    val coverage: TranslationCoverage = TranslationCoverage(),
)

@Serializable
data class CourseTranslationListItem(
    val itemId: String = "",
    val itemType: String = "",
    val title: String = "",
    val body: String = "",
    val hasPublished: Boolean? = null,
    val hasDraft: Boolean? = null,
    val targetLocale: String? = null,
    val translatedTitle: String? = null,
    val translatedBody: String? = null,
    val isDraft: Boolean? = null,
    val machineTranslationDraft: Boolean? = null,
    val publishedAt: String? = null,
    val version: Long? = null,
    val glossaryMatches: List<GlossaryMatchSpan> = emptyList(),
)

@Serializable
data class GlossaryMatchSpan(
    val sourceTerm: String = "",
    val targetTerm: String = "",
    val start: Int? = null,
    val end: Int? = null,
)

@Serializable
data class CourseGlossaryListResponse(
    val entries: List<CourseGlossaryEntry> = emptyList(),
)

@Serializable
data class CourseGlossaryEntry(
    val id: String = "",
    val sourceTerm: String = "",
    val targetTerm: String = "",
    val sourceLocale: String? = null,
    val targetLocale: String? = null,
)

@Serializable
data class TranslationLocalesResponse(
    val locales: List<TranslationCoverage> = emptyList(),
)

@Serializable
data class SaveCourseTranslationBody(
    val targetLocale: String,
    val sourceLocale: String? = null,
    val translatedTitle: String? = null,
    val translatedBody: String? = null,
    val isDraft: Boolean? = null,
    val machineTranslationDraft: Boolean? = null,
    val version: Long? = null,
)

@Serializable
data class PublishCourseTranslationBody(
    val targetLocale: String,
)

@Serializable
data class AddGlossaryEntryBody(
    val sourceTerm: String,
    val targetTerm: String,
    val targetLocale: String,
    val sourceLocale: String? = "en",
)

File: clients/ios/Lextures/Core/LMS/LMSFeatureModelsCourseTranslations.swift
import Foundation

/// Course translation / glossary models (M13.9).
struct TranslationCoverage: Codable, Hashable, Identifiable {
    var targetLocale: String
    var totalItems: Int
    var translatedItems: Int
    var percent: Double
    var untranslated: [TranslatableCourseItem]?

    var id: String { targetLocale }

    var percentInt: Int {
        Int(percent.rounded())
    }
}

struct TranslatableCourseItem: Codable, Hashable, Identifiable {
    var itemId: String
    var itemType: String
    var title: String
    var body: String
    var hasPublished: Bool?
    var hasDraft: Bool?

    var id: String { itemId }
}

struct CourseTranslationListResponse: Codable, Hashable {
    var items: [CourseTranslationListItem]
    var coverage: TranslationCoverage
}

struct CourseTranslationListItem: Codable, Hashable, Identifiable {
    var itemId: String
    var itemType: String
    var title: String
    var body: String
    var hasPublished: Bool?
    var hasDraft: Bool?
    var targetLocale: String?
    var translatedTitle: String?
    var translatedBody: String?
    var isDraft: Bool?
    var machineTranslationDraft: Bool?
    var publishedAt: String?
    var version: Int64?
    var glossaryMatches: [GlossaryMatchSpan]?

    var id: String { itemId }
}

struct GlossaryMatchSpan: Codable, Hashable {
    var sourceTerm: String
    var targetTerm: String
    var start: Int?
    var end: Int?
}

struct CourseGlossaryListResponse: Codable, Hashable {
    var entries: [CourseGlossaryEntry]
}

struct CourseGlossaryEntry: Codable, Hashable, Identifiable {
    var id: String
    var sourceTerm: String
    var targetTerm: String
    var sourceLocale: String?
    var targetLocale: String?
}

struct TranslationLocalesResponse: Codable, Hashable {
    var locales: [TranslationCoverage]
}

struct SaveCourseTranslationBody: Codable, Hashable {
    var targetLocale: String
    var sourceLocale: String?
    var translatedTitle: String?
    var translatedBody: String?
    var isDraft: Bool?
    var machineTranslationDraft: Bool?
    var version: Int64?
}

struct PublishCourseTranslationBody: Codable, Hashable {
    var targetLocale: String
}

struct AddGlossaryEntryBody: Codable, Hashable {
    var sourceTerm: String
    var targetTerm: String
    var targetLocale: String
    var sourceLocale: String?
}

